text/0005-disallow-whitespace-in-entityuid.md

```diff
@@ -0,0 +1,98 @@
+- Date Landed:
+
+## Summary
+
+Disallow embedded whitespace, comments, and control characters in string representations of the EntityUID
+
+## Basic example
+Currently, the following syntax is legal:
+```
+EntityTypeName::from_str("ExampleCo  ::  Photoflash  ::  User  //comment");
+```
+This change would modify the function so that the string input must be normalized; i.e. all extraneous whitespace, control characters, and comments removed.
+```
+EntityTypeName::from_str("ExampleCo::Photoflash::User");
+```
+
+## Motivation
+Similar to other programming languages such as Rust, Cedar is currently whitespace agnostic. This leads to the ability to embed whitespace, control characters like newlines, and comments around the `::` delimiter in an Entity UID.
+
+For example, the following syntax is valid:
+```
+permit( 
+  principal == ExampleCo :: Photoflash  ::  //This is a comment
+       :: User::"alice",
+  action,
+  resource
+);
+
+permit( 
+  principal == ExampleCo::Photoflash::User:://comment
+
+  "alice",
+  action,
+  resource
+);
+```
+
+This capability was little known, even amongst many Cedar team members, and was only identified as a result of a typo in a Cedar demo. While this behavior may be OK inside policy statements, there is a risk when policy contents such as the EntityTypeName are used as strings outside the context of policy statements, such as in the JSON schema representation. For laypersons who are not Cedar experts, these values appear to be normalized strings rather than fragments of Cedar policy syntax, leading to bugs in application code when whitespace and comments can appear in the value.
+
+Examples:
+1. The Cedar Schema format models the schema configuration under a JSON key for the namespace. Policy stores which index schemas by namespace are unlikely to recognize the need to normalize the value, leading to the possibility of storing duplicate schema definitions for "ExampleCo::Photoflash" and "ExampleCo  ::  Photoflash" and indeterminate behavior regarding which schema takes effect at runtime.
+2. Policy Stores can implement logic that relies on string comparisons against the EntityTypeName. In a real issue, an application using Cedar sought to preclude callers from passing Actions in the inline slice of entity definitions. It did so by checking if an EntityTypeName matched `.*::Action`. It presumed that `:: Action` was invalid syntax and would be rejected by the Cedar evalator, the same as any other syntatically invalid input. This resulted in a bug, as it allowed callers to bypass the extra validation that the application sought to enforce.
+3. Customers are anticipated to build meta-permissions layers that restrict callers to manipulating policy store contents for only certain namespaces. This may lead to policies such as `forbid(...) when {context.namespace = "ExampleCo::Photoflash"};`. There is a risk that an unauthorized actor could bypass this restriction by using a namespace with embedded spaces. 
```

Comment:
The example doesn't rely on Cedar being able to extract the namespace from an entity. I think this is where Darin is coming from:

The `context.namespace` would be populated by the application submitting the Cedar request, and the action (not shown) would relate to adding a policy to the store (say). The concern is that `context.namespace` is populated in part by user data. The application would want to make sure the user data is a legit namespace, so it would run it through the Cedar parser. What if the user data was `"ExampleCo::  Photoflash"`? The parser would _accept_ it as of right now. That checked, the application needs to _normalize_ this data, so that it stores `"ExampleCo::Photoflash"` in `context.namespace`, not `"ExampleCo::  Photoflash"`. This way, it matches the `when` condition as expected. With the change proposed in this RFC, the normalization step is not required, which means the application code will be less error prone.

Reply:
Gotcha, I understand now.